test/stability-tests/src/it/java/software/amazon/awssdk/stability/tests/utils/StabilityTestRunner.java

```diff
@@ -0,0 +1,275 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.stability.tests.utils;
+
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.IntFunction;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.core.exception.SdkServiceException;
+import software.amazon.awssdk.stability.tests.ExceptionCounter;
+import software.amazon.awssdk.stability.tests.TestResult;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * Stability tests runner.
+ *
+ * There are two ways to run the tests:
+ *
+ * - providing futureFactories requestCountPerRun and totalRuns. eg:
+ *
+ * StabilityTestRunner.newRunner()
+ *                    .futureFactory(i -> CompletableFuture.runAsync(() -> LOGGER.info(() ->
+ *                      "hello world " + i), executors))
+ *                    .testName("test")
+ *                    .requestCountPerRun(10)
+ *                    .delaysBetweenEachRun(Duration.ofMillis(1000))
+ *                    .totalRuns(5)
+ *                    .run();
+ *
+ * The tests runner will create futures from the factories and run the 10 requests per run for 5 runs with 1s delay between
+ * each run.
+ *
+ * - providing futures, eg:
+ *
+ *         StabilityTestRunner.newRunner()
+ *                            .futures(futures)
+ *                            .testName("test")
+ *                            .run();
+ * The tests runner will run the given futures in one run.
+ */
+public class StabilityTestRunner {
+
+    private static final Logger log = Logger.loggerFor(StabilityTestRunner.class);
+    private static final double ALLOWED_FAILURE_RATIO = 0.05;
```

Comment:
How did we determine this number? What are the expected exceptions?

Reply:
It's in the design doc, will ping you the link. The expected exceptions are `SdkServiceException`s, `IOException`s